test/caching_test.rb

```diff
@@ -0,0 +1,61 @@
+require File.expand_path('../support/test_helper', __FILE__)
+
+class CachingTestTest < Minitest::Test
+  URI = "http://example.com/test-schema.json"
+
+  def test_caching
+    initial_schema = {
+      "type" => "string"
+    }
+    schema = JSON::Schema.new(initial_schema, URI)
+    JSON::Validator.add_schema(schema)
+    assert_valid(URI, "foo", :clear_cache => false)
+
+    initial_schema = {
+      "type" => "number"
+    }
+    schema = JSON::Schema.new(initial_schema, URI)
+    JSON::Validator.add_schema(schema)
+    assert_valid(URI, "foo")
+  end
+
+  def test_clear_cache
+    initial_schema = {
+      "type" => "string"
+    }
+    schema = JSON::Schema.new(initial_schema, URI)
+    JSON::Validator.add_schema(schema)
+    assert_valid(URI, "foo", :clear_cache => true)
+
+    initial_schema = {
+      "type" => "number"
+    }
+    schema = JSON::Schema.new(initial_schema, URI)
+    JSON::Validator.add_schema(schema)
+    assert_valid(URI, 123)
+  end
+
+  def test_cache_schemas
+    suppress_warnings do
+      JSON::Validator.cache_schemas = true
+    end
+
+    initial_schema = {
+      "type" => "string"
+    }
+    schema = JSON::Schema.new(initial_schema, URI)
+    JSON::Validator.add_schema(schema)
+    assert_valid(URI, "foo")
+
+    initial_schema = {
+      "type" => "number"
+    }
+    schema = JSON::Schema.new(initial_schema, URI)
+    JSON::Validator.add_schema(schema)
+    assert_valid(URI, 123)
```

Comment:
The logic here is wrong - when cache_schemas is true this test should fail because the previous one should still be cached